src/Columns/ColumnSparse.cpp

```diff
@@ -0,0 +1,762 @@
+
+ColumnPtr ColumnSparse::index(const IColumn & indexes, size_t limit) const
+{
+    return selectIndexImpl(*this, indexes, limit);
+}
+
+template <typename Type>
+ColumnPtr ColumnSparse::indexImpl(const PaddedPODArray<Type> & indexes, size_t limit) const
+{
+    limit = limit ? std::min(limit, indexes.size()) : indexes.size();
+
+    auto res_offsets = offsets->cloneEmpty();
+    auto & res_offsets_data = assert_cast<ColumnUInt64 &>(*res_offsets).getData();
+    auto res_values = values->cloneEmpty();
+    res_values->insertDefault();
+
+    for (size_t i = 0; i < limit; ++i)
+    {
+        size_t index = getValueIndex(indexes[i]);
+        if (index != 0)
+        {
+            res_values->insertFrom(*values, index);
+            res_offsets_data.push_back(i);
+        }
+    }
+
+    return ColumnSparse::create(std::move(res_values), std::move(res_offsets), limit);
+}
+
+int ColumnSparse::compareAt(size_t n, size_t m, const IColumn & rhs_, int null_direction_hint) const
+{
+    if (const auto * rhs_sparse = typeid_cast<const ColumnSparse *>(&rhs_))
+        return values->compareAt(getValueIndex(n), rhs_sparse->getValueIndex(m), rhs_sparse->getValuesColumn(), null_direction_hint);
+
+    return values->compareAt(getValueIndex(n), m, rhs_, null_direction_hint);
+}
+
+void ColumnSparse::compareColumn(const IColumn & rhs, size_t rhs_row_num,
+                    PaddedPODArray<UInt64> * row_indexes, PaddedPODArray<Int8> & compare_results,
+                    int direction, int nan_direction_hint) const
+{
+    if (row_indexes)
+    {
+        /// TODO: implement without conversion to full column.
+        auto this_full = convertToFullColumnIfSparse();
+        auto rhs_full = rhs.convertToFullColumnIfSparse();
+        this_full->compareColumn(*rhs_full, rhs_row_num, row_indexes, compare_results, direction, nan_direction_hint);
+    }
+    else
+    {
+        const auto & rhs_sparse = assert_cast<const ColumnSparse &>(rhs);
+        PaddedPODArray<Int8> nested_result;
+        values->compareColumn(rhs_sparse.getValuesColumn(), rhs_sparse.getValueIndex(rhs_row_num),
+            nullptr, nested_result, direction, nan_direction_hint);
+
+        const auto & offsets_data = getOffsetsData();
+        compare_results.resize_fill(_size, nested_result[0]);
+        for (size_t i = 0; i < offsets_data.size(); ++i)
+            compare_results[offsets_data[i]] = nested_result[i + 1];
+    }
+}
+
+int ColumnSparse::compareAtWithCollation(size_t n, size_t m, const IColumn & rhs, int null_direction_hint, const Collator & collator) const
+{
+    if (const auto * rhs_sparse = typeid_cast<const ColumnSparse *>(&rhs))
+        return values->compareAtWithCollation(getValueIndex(n), rhs_sparse->getValueIndex(m), rhs_sparse->getValuesColumn(), null_direction_hint, collator);
+
+    return values->compareAtWithCollation(getValueIndex(n), m, rhs, null_direction_hint, collator);
+}
+
+bool ColumnSparse::hasEqualValues() const
+{
+    return offsets->empty();
```

Comment:
looks like we also need to check if values has equal values too, even if offsets are empty
also I asked amos why do we need this method at all - it is still not used

Reply:
And we need to check the case with only defaults.
(as I see, it is possible that values contains default not only at 0 pos)

Reply:
Implemented a correct implementation for all cases.